refactor(cache): optimize caching and dispatching behaviour

Damego opened the pull request:
## About

This pull request:
1. Moves getting object ID into separated method.
2. Adds method to get object IDs. It needs for `message_delete_bulk`, `emojis_update` and `stickers_update` events. It allows properly cache emojis and stickers.
3. Caches `GuildMember` like `Member`
4. Deletes messages from cache on `message_delete_bulk` event.
5. Moves `__modify_guild_cache` to own method.
6. Makes properly cache `Guild.members` and `Guild.roles`.
7. Properly dispatches remove/delete event so we no need to use raw remove/delete event.
8. Dispatches `_update` event if there are no id and ids. It allows use `on_channel_pins_update`, `on_webhooks_update` events without `raw` prefix.

## Checklist

- [x] I've ran `pre-commit` to format and lint the change(s) made.
- [x] I've checked to make sure the change(s) work on `3.8.6` and higher.
- [ ] This fixes/solves an [Issue](https://github.com/goverfl0w/discord-interactions/issues) (If existent):.
  - resolves #
- I've made this pull request for/as: (check all that apply)
  - [x] Documentation
  - [ ] Breaking change
  - [x] New feature/enhancement
  - [ ] Bugfix

Damego committed:
refactor: just refactor

1. Moved getting object ID into separated method.
2. Add method to get object ids. Needs for `message_delete_bulk`, `emoji_update` and `stickers_update`
3. Cache `GuildMember` like `Member`
4. Deleting messages from cache on `message_delete_bulk` event

Damego committed:
chore: run pre-commit

Damego committed:
refactor: merge conditions into one

sillyElvy reviewed (changes requested):
Add emojis to global cache in the guild obj

Damego committed:
refactor: add emojis to guild cache in attrs_post_init

Damego committed:
fix: typo

Damego committed:
refactor: move message cache control to method

- Remove message from cache on `message_delete` event

pre-commit-ci[bot] committed:
ci: correct from checks.

Damego committed:
fix the fix of bug

Damego committed:
docs: capitalize return and ` to '

Damego committed:
refactor: leave only `message`

Catalyst4222 commented on interactions/api/gateway/client.py:
```diff
@@ -441,55 +443,35 @@ def _dispatch_event(self, event: str, data: dict) -> None:
-                            id = getattr(obj, f"{model_name}_id", None)
+                    ids = self.__get_object_ids(obj, model)
 
                 def __modify_guild_cache():
```
While you're optimizing, can you move `__modify_guild_cache` into its own function?

Damego replied:
Sure

Damego replied:
Done

sillyElvy commented on interactions/api/models/guild.py:
```diff
@@ -360,6 +360,8 @@ def __attrs_post_init__(self):  # sourcery skip: last-if-guard
                     self.member_count = guild.member_count
                 if not self.presences:
                     self.presences = guild.presences
+                if not self.emojis:
+                    self.emojis = guild.emojis
```
Please also add this to adding it to the global cache as above done with other objects

Damego replied:
Emojis doesn't exists in global cache.

sillyElvy replied:
Why would they not?

Damego replied:
> Why would they not? 

Discord has only one event with emoji. Its `guild_emojis_update`. On unstable it doesn't gets emojis and not stores them in cache.
To cache emojis we should get emojis from `GuildEmojis` object and do some stuff with cache. Same for stickers.

Damego committed:
ref: move `__modify_guild_cache` to own method

pre-commit-ci[bot] committed:
ci: correct from checks.

Damego committed:
docs: add and update docstrings

Damego committed:
refactor: properly caching members and roles

pre-commit-ci[bot] committed:
ci: correct from checks.

Damego commented on interactions/api/gateway/client.py:
```diff
@@ -435,83 +438,51 @@ def _dispatch_event(self, event: str, data: dict) -> None:
+                # TODO: Move to own method?
+                guild_obj = None
+                if guild_model is Role:
+                    guild_obj = Role(**data["role"]) if "role" in data else _guild_cache.get(id)
```
`guild_role_delete` doesn't contain `role` in data so i took cached state

Damego committed:
refactor: refuse from raw remove events and ...

put `obj` to `__modify_guild_cache` if guild_obj is None

pre-commit-ci[bot] committed:
ci: correct from checks.

Damego committed:
fix: `guild_obj` referenced before assignment

Damego committed:
Merge branch 'refactor-gateway-caching' of https://github.com/Damego/library into refactor-gateway-caching

Damego committed:
fix: help me

There are only two models which can be converted to model without `Guild` prefix.

Damego commented:
Looks like it works fine for me

sillyElvy commented on interactions/api/gateway/client.py:
```diff
@@ -431,87 +434,61 @@ def _dispatch_event(self, event: str, data: dict) -> None:
+                # I don't like this but idk what i should do then
+                guild_obj = guild_model = None
+                if model is GuildRole:
+                    guild_obj = Role(**data["role"]) if "role" in data else None
```
```suggestion
                    guild_obj = Role(**role_data) if (role_data := data.get("role")) else None
```

Damego replied:
Then i suggest use walrus operator

sillyElvy replied:
Idek if that works there

Damego replied:
This works
```py
guild_obj = Role(**role_data) if (role_data := data.get("role")) else None
```

sillyElvy replied:
Modified suggestion

Damego committed:
refactor: use walrus operator

Co-authored-by: EdVraz <88881326+EdVraz@users.noreply.github.com>

Damego committed:
fix: add missed obj to dispatch

Damego committed:
refactor: remove extra dispatch

Damego committed:
refactor: message already removed from cache?

Damego committed:
refactor: remove `__delete_message_cache`

Damego committed:
refactor: dispatch `guild_emojis_update` without `raw`

pre-commit-ci[bot] committed:
ci: correct from checks.

Damego committed:
refactor: change condition for `ids`

Damego committed:
Merge branch 'refactor-gateway-caching' of https://github.com/Damego/library into refactor-gateway-caching

Damego committed:
refactor: change condition in getting ids

AstreaTSS reviewed (approved):
Haven't verified the code works, but it looks solid.

Damego committed:
feat: add support for `_execution` events

Damego committed:
refactor: dispatch update event if there are no id & ids

Damego commented:
New changes:
- ~~Add support for `_execution` event.~~ No need lol. I forgot we have `else` block
- Dispatches `_update` event if there are no id and ids. It allows use `on_channel_pins_update`, `on_webhooks_update` events without `raw` prefix.

Damego committed:
revert: add support for `_execution` event

Damego committed:
refactor: some stuff

Damego committed:
refactor: use `cache.merge` to resolve guild overloading
